hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java

```diff
@@ -0,0 +1,191 @@
+                op.hasSupplyKey(), op.supplyKey(),
+                op.hasFreezeKey(), op.freezeKey(),
+                op.hasFeeScheduleKey(), op.feeScheduleKey(),
+                op.hasPauseKey(), op.pauseKey());
+        // validate custom fees length
+        validateTrue(
+                op.customFeesOrElse(emptyList()).size() <= config.maxCustomFeesAllowed(), CUSTOM_FEES_LIST_TOO_LONG);
+    }
+
+    /**
+     * Validates initial supply and decimals based on token type
+     * @param type token type
+     * @param initialSupply initial supply
+     * @param decimals decimals
+     * @throws PreCheckException if validation fails
+     */
+    private void validateTokenType(final TokenType type, final long initialSupply, final int decimals)
+            throws PreCheckException {
+        validateTruePreCheck(type == FUNGIBLE_COMMON || type == NON_FUNGIBLE_UNIQUE, NOT_SUPPORTED);
+        if (type == FUNGIBLE_COMMON) {
+            validateTruePreCheck(initialSupply >= 0, INVALID_TOKEN_INITIAL_SUPPLY);
+            validateTruePreCheck(decimals >= 0, INVALID_TOKEN_DECIMALS);
+        } else {
+            validateTruePreCheck(initialSupply == 0, INVALID_TOKEN_INITIAL_SUPPLY);
+            validateTruePreCheck(decimals == 0, INVALID_TOKEN_DECIMALS);
+        }
+    }
+
+    /**
+     * Validates supply type and max supply
+     * @param supplyType supply type
+     * @param maxSupply max supply
+     * @throws PreCheckException if validation fails
+     */
+    private void validateSupplyType(final TokenSupplyType supplyType, final long maxSupply) throws PreCheckException {
+        validateTruePreCheck(supplyType == INFINITE || supplyType == FINITE, NOT_SUPPORTED);
+        if (supplyType == INFINITE) {
+            validateTruePreCheck(maxSupply == 0, INVALID_TOKEN_MAX_SUPPLY);
+        } else {
+            validateTruePreCheck(maxSupply > 0, INVALID_TOKEN_MAX_SUPPLY);
+        }
+    }
+
+    /**
+     * Validates if the token and account already have relationship and if the account has reached the limit of
+     * associations.
+     * These checks need to be done before the token is created and associated to treasury or any custom
+     * fee collector accounts.
+     * @param entitiesConfig entities config
+     * @param tokensConfig tokens config
+     * @param account account to associate with
+     * @param token token to associate with
+     * @param tokenRelStore token relation store
+     */
+    public void validateAssociation(
+            @NonNull final EntitiesConfig entitiesConfig,
+            @NonNull final TokensConfig tokensConfig,
+            @NonNull final Account account,
+            @NonNull final Token token,
+            @NonNull final WritableTokenRelationStore tokenRelStore) {
+        validateFalse(
+                entitiesConfig.limitTokenAssociations()
+                        && account.numberAssociations() + 1 > tokensConfig.maxPerAccount(),
+                TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
```

Comment:
Again, seems there should be a check for tokens per account _and_ max tokens overall

Reply:
Max tokens created check is in start of handle and will throw `MAX_ENTITIES_IN_PRICE_REGIME_HAVE_BEEN_CREATED`